p2p/channel.go

```diff
@@ -80,86 +81,218 @@ type stream struct {
+	// get updated peer address.
+	proxyConn *net.UDPConn
 }
 
 // channel implements Channel interface.
 type channel struct {
 	mu   sync.RWMutex
 	once sync.Once
 
-	tr               *transport
-	serviceConn      *net.UDPConn
-	topicHandlers    map[string]HandlerFunc
-	streams          map[uint64]*stream
-	nextStreamID     uint64
-	privateKey       PrivateKey
-	peerPubKey       PublicKey
-	peerAddr         *net.UDPAddr
-	localAddr        *net.UDPAddr
-	blockCrypt       kcp.BlockCrypt
-	stop             chan struct{}
-	sendQueue        chan *transportMsg
+	// tr is transport containing network related connections for p2p to work.
+	tr *transport
+
+	// serviceConn is separate connection which is created outside of p2p channel when
+	// performing initial NAT hole punching or manual conn. It is here just because it's more easy
+	// to pass it to services as p2p channel will be available anyway.
+	serviceConn *net.UDPConn
+
+	// topicHandlers is similar to HTTP Server handlers and is responsible for handling peer requests.
+	topicHandlers map[string]HandlerFunc
+
+	// streams is temp map to create request/response pipelines. Each stream is created on send and contains
+	// channel to which receive loop should eventually send peer reply.
+	streams      map[uint64]*stream
+	nextStreamID uint64
+
+	// privateKey is channel's private key. For now it's here just to be able to recreate the same channel for unit tests.
+	privateKey PrivateKey
+
+	// peer is remote peer holding it's public key and address.
+	peer *peer
+
+	// localSessionAddr is KCP UDP conn address to which packets are written from remote conn.
+	localSessionAddr *net.UDPAddr
+
+	// sendQueue is a queue to which channel puts messages for sending. Message is not send directly to remote peer
+	// but to proxy conn which is when responsible for sending to remote
+	sendQueue chan *transportMsg
+
+	// upnpPortsRelease should be called to close mapped upnp ports when channel is closed.
 	upnpPortsRelease []func()
+
+	// stop is used to stop all running goroutines.
+	stop chan struct{}
 }
 
 // newChannel creates new p2p channel with initialized crypto primitives for data encryption
 // and starts listening for connections.
-func newChannel(punchedConn *net.UDPConn, privateKey PrivateKey, peerPubKey PublicKey) (*channel, error) {
-	localPort := punchedConn.LocalAddr().(*net.UDPAddr).Port
-	remotePort := punchedConn.RemoteAddr().(*net.UDPAddr).Port
-	log.Debug().Msgf("Creating p2p channel with local port: %d, remote port: %d", localPort, remotePort)
+func newChannel(remoteConn *net.UDPConn, privateKey PrivateKey, peerPubKey PublicKey) (*channel, error) {
+	localAddr := remoteConn.LocalAddr().(*net.UDPAddr)
+	peerAddr := remoteConn.RemoteAddr().(*net.UDPAddr)
+	log.Debug().Msgf("Creating p2p channel with local port: %d, remote port: %d", localAddr.Port, peerAddr.Port)
 
-	blockCrypt, err := newBlockCrypt(privateKey, peerPubKey)
+	remoteConn.Close()
```

Comment:
This looks a bit awkward. We pass alive conn and close right away. Would extract this close() to other func `remoteConn = updateConn(localAddr)`